Relax the relationship of RateLimit headers with Retry-After and add examples (#73)
ioggstream/draft-polli-ratelimit-headers: draft-polli-ratelimit-headers.md

of the response status code.  This includes throttled responses.

If a response contains both the `Retry-After` and the `RateLimit-Reset` header fields,

> the value of `RateLimit-Reset` MAY reference the same point in time as

Reviewer: ```suggestion
the value of `RateLimit-Reset` SHOULD reference the same point in time as
```

Reviewer: If you agree, just commit suggestion and resolve conversations.

> `Retry-After`.

Reviewer: If we do not mandate any relation, I'd remove the whole sentence.

Author: I agree, but It could appear to someone reading this that we forgot about the `Retry-After` timestamp when both headers are present. IOW, it is asserting we've looked into it and we add no requirement.

Reviewer: Let's `SHOULD` then :P

Author: resolved :+1:

A client exhausted its quota and the server throttles the request
sending the `Retry-After` response header field.

> The values of `Retry-After` and `RateLimit-Reset` reference the same moment,

Reviewer: ```suggestion
In this example, the values of `Retry-After` and `RateLimit-Reset` reference the same moment,
```

> but there is no requirement that they do so.

Reviewer: ```suggestion
but this is not a requirement.
```

A client exhausted its quota and the server throttles the request
sending the `Retry-After` response header field.

> In this example, the values of `Retry-After` and `RateLimit-Reset` reference the same moment,

Reviewer: +1

sending the `Retry-After` response header field.

The values of `Retry-After` and `RateLimit-Reset` reference the same moment,

> but this is not a requirement.

Reviewer: +1

The policy conveyed by `RateLimit-Limit` states that
the server accepts 100 quota-units per minute.

> Due to resource scarcity, the server artificially lowers

Reviewer: ```suggestion
To avoid resource exhaustion, the server artificially lowers
```

the actual limits returned in the throttling headers.

> The current policy then advertises

Reviewer: ```suggestion
The  `RateLimit-Remaining` then advertises
```

The policy conveyed by `RateLimit-Limit` states that
the server accepts 100 quota-units per minute.

> To avoid resource exhaustion, the server artificially lowers

Reviewer: +1

{"hello": "world"}
~~~

> ### Dynamic limits for pushing back and slowing down

Reviewer: I understand the goal for this example: just slow down without waiting.

In case of 429/503 with 

```
ratelimit-remaining: 0
ratelimit-reset:  > 0
```

the semantic suggests to wait for ratelimit-reset though.

I'm not sure we should advertise `rl-limit: 0`. Which is the intended behavior.

```
ratelimit-limit: 0, 15;w=20 
```

Author: The remaining units being `0` is actually one of two possible ways to slow down clients. This one is forcing them to stop and wait a little, so it acts as a hard limit at that point (need to wait), whereas the previous example (which this one follows up on) shows a case in which the server lowers an initial window of 60 seconds with 100 units halfway through to considerably reduce the available units to 10 (9 after the request is responded) so that the client knows to use them during the rest of that window, which reduces the average request rate over that period but still accepts consuming all the remaining units right away. In this latter case the server might just decide the spike would be unacceptable and tell the client to wait a little, like the example shown here:

"Wait for 20 seconds without requests, then the next window is going to have only 15 units for 20 seconds". Presumably once the resource starvation is over the server could carry on with the previous limits, returning to a windows of 60 seconds with 100 units.

Continuing the previous example, let's say the client waits 10 seconds and
performs a new request which, due to resource exhaustion, the server rejects
and pushes back, advertising `RateLimit-Remaining: 0` for the next 20 seconds.

> The server decides to advertise a new smaller window with a lower limit to slow

Reviewer: ```suggestion
The server advertises a smaller window with a lower limit to slow
```

down the client for the rest of its original window after the 20 seconds elapse.

Request:

~~~
GET /items/123

~~~

Response:

~~~
HTTP/1.1 429 Too Many Requests

> RateLimit-Limit: 0, 15;w=20

Reviewer: why sending `ratelimit-limit: 0` ?

Author: See https://github.com/ioggstream/draft-polli-ratelimit-headers/pull/73#discussion_r375361958

Reviewer: I think that should be done via:
```
ratelimit-reset: 0
```

In any case, as we are focusing on the semantic of the headers, even this behavior is legitimate.

Ratelimit-Remaining: 0
Ratelimit-Reset: 20

> 

Reviewer: Add a payload like the following containing explanations. 

```
{
  "status": 429,
  "detail": "Just slow down without waiting."
}
```

Author: Sounds good, will do. :+1:

~~~

## Dynamic limits for pushing back with Retry-After and slow down

Alternatively, given the same context where the previous example starts, we
can convey the same information to the client via the Retry-After header, with
the advantage that the server can now specify the policy's nominal limit and
window that will apply after the reset, ie. assuming the resource exhaustion
is likely to be gone by then, so the advertised policy does not need to be
adjusted, yet we managed to stop requests for a while and slow down the rest of
the current window.

Request:

~~~
GET /items/123

~~~

Response:

~~~
HTTP/1.1 429 Too Many Requests
Retry-After: 20
RateLimit-Limit: 15, 100;w=60
Ratelimit-Remaining: 15
Ratelimit-Reset: 40

> 

Reviewer: Add a payload like the following containing explanations. 

```
{
  "status": 429,
  "detail": "Wait 20 seconds, then slow down!"
}
```

Author: :+1:

performs a new request which, due to resource exhaustion, the server rejects
and pushes back, advertising `RateLimit-Remaining: 0` for the next 20 seconds.

> The server advertises a smaller window with a lower limit to slow

Reviewer: +1

   in case of resource saturation (eg. thrashing, connection queues too long,
   Service Level Objectives not meet, ..).

   Availability can be improved by dynamically lowering the values returned by

>    the rate-limit headers to slow down clients, and retry-after can be used to

Reviewer: ```suggestion
   the `RateLimit-*` headers to slow down clients, and `Retry-After` can be used to
```
